### PR TITLE
pkg/sqlutil: add QueryHook for wrapping Queryers with timeouts, prometheus metrics and slow query logging

### PR DESCRIPTION
**Pull request by jmank88**
This PR introduces `sqlutil.QueryHook`s based on the core `pq.Q` type's behavior: https://github.com/smartcontractkit/chainlink/blob/5f212bbc8485fc1fb20076ef5df7842f7bbfc6c1/core/services/pg/q.go#L117

https://smartcontract-it.atlassian.net/browse/BCF-2978

image
image

### CODE REVIEW

File: pkg/sqlutil/log.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+var promSQLQueryTime = promauto.NewHistogram(prometheus.HistogramOpts{
+	Name:    "sql_query_timeout_percent",
+	Help:    "SQL query time as a percentage of timeout.",
+	Buckets: []float64{10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120},
```
Comment (patrickhuie19) [inline pkg/sqlutil/log.go:24]: For my understanding: why stop/end at 120?

Comment (jmank88) [reply pkg/sqlutil/log.go:24]: The response to context cancellation is generally very fast, so we don't see values beyond this point very often :shrug: 
Or at least we didn't when I calibrated this scale for the original one that I copied here. Could double check in the grafana stats.

File: pkg/sqlutil/log.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
+package sqlutil
```
Comment (patrickhuie19) [inline pkg/sqlutil/log.go:1]: This PR made me think of `core/services/pg/sqlx.go` - did you imagine we would consolidate the `Queryer` interface there into the one in this package?

Comment (jmank88) [reply pkg/sqlutil/log.go:1]: Yeah, this is meant to replace it.

**Review by samsondav — APPROVED** [review]

**Review by patrickhuie19 — DISMISSED** [review]

File: pkg/sqlutil/interceptor.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	"context"
+	"database/sql"
+
+	"github.com/jmoiron/sqlx"
+)
+
+var _ Queryer = &InterceptedQueryer{}
+
+type Interceptor func(ctx context.Context, do func(context.Context) error, query string, args ...any) error
```
Comment (patrickhuie19) [inline pkg/sqlutil/interceptor.go:12]: Does the `database/sql` package or `sqlx` package provide an Interceptor type + APIs we can use? Seems like we aren't the only ones who would want this functionality as well.

Comment (jmank88) [reply pkg/sqlutil/interceptor.go:12]: I had the same thought but couldn't find one for sqlx

**Review by DylanTinianov — DISMISSED** [review]

File: pkg/sqlutil/hook.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+// HookingQueryer is a Queryer which invokes a QueryHook on each call.
+type HookingQueryer struct {
+	q    Queryer
+	lggr logger.Logger
+	hook QueryHook
+}
+
+// An QueryHook is a func that is executed for each query, providing an opportunity to measure, log, inspect/modify errors, etc.
+// The do func *must* be called.
+// Logs emitted through the provided logger.Logger will have the caller's line info.
```
Comment (jmank88) [inline pkg/sqlutil/hook.go:23]: I ended up adding a logger to get  the caller info right.

**Comment by jmank88** [comment]
Updated naming, added tests for caller line info.

File: pkg/sqlutil/monitor.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+const slowMsg = "SLOW SQL QUERY"
+
+var promSQLQueryTime = promauto.NewHistogram(prometheus.HistogramOpts{
+	Name:    "sql_query_timeout_percent",
```
Comment (jmank88) [inline pkg/sqlutil/monitor.go:23]: I don't think we will be able to swap this in for everything at once in core, so we'll have to find a way to share this prom metric during the transition. Maybe temporarily exporting is the simplest thing to do :thinking:

File: pkg/sqlutil/hook.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+// NewHookingQueryer returns a new [HookingQueryer] that calls each [QueryHook] in the provided order.
+func NewHookingQueryer(q Queryer, l logger.Logger, hs ...QueryHook) *HookingQueryer {
+	iq := HookingQueryer{q: q,
+		lggr: logger.Helper(l, 2), // skip our own wrapper and one interceptor
+		hook: noopHook,
+	}
+	switch len(hs) {
+	case 0:
+		return &iq
+	case 1:
```
Comment (patrickhuie19) [inline pkg/sqlutil/hook.go:37]: /nit `case default` ?

**Review by samsondav — DISMISSED** [review]

File: pkg/sqlutil/hook.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		iq.hook = hs[0]
+		return &iq
+	}
+
+	for i := len(hs) - 1; i >= 0; i-- {
```
Comment (krehermann) [inline pkg/sqlutil/hook.go:42]: please add a comment to guide future readers. i can't figure out what this is doing after 3 minutes of reading it.

Comment (jmank88) [reply pkg/sqlutil/hook.go:42]: Will do. It is chaining them by exploiting the `do func` that they are already expected to call. The tricky part from getting the logger skip levels right.

File: pkg/sqlutil/hook_test.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	getDur = 10 * time.Millisecond
+	selDur = 200 * time.Millisecond
+)
+
+func TestNewInterceptedQueryer(t *testing.T) {
+	lggr, ol := logger.TestObserved(t, zapcore.InfoLevel)
+	var q Queryer = &queryer{}
+	var sentinelErr = errors.New("intercepted error")
+	const fakeError = "fake warning"
+	q = NewHookingQueryer(q, lggr, TimeoutHook(selDur/2), noopHook, MonitorHook(func() bool { return true }), noopHook, func(ctx context.Context, lggr logger.Logger, do func(context.Context) error, query string, args ...any) error {
```
Comment (krehermann) [inline pkg/sqlutil/hook_test.go:31]: does this test validate the order of the hook executions? if so, can you tell me where that is happening?

Comment (jmank88) [reply pkg/sqlutil/hook_test.go:31]: It does insofar as it wouldn't work the same if they were reversed, but I'll try to find a way to make it more clear.

**Review by DylanTinianov — DISMISSED** [review]
Looks good

**Review by samsondav — APPROVED** [review]

File: pkg/sqlutil/monitor.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+// PromSQLQueryTime is exported temporarily while transitioning the core ORMs.
+var PromSQLQueryTime = promauto.NewHistogram(prometheus.HistogramOpts{
+	Name:    "sql_query_timeout_percent",
+	Help:    "SQL query time as a percentage of timeout.",
+	Buckets: []float64{10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120},
+})
+
+// MonitorHook returns a [QueryHook] that measures the timing of each query and logs about slow queries at increasing levels of severity.
+// When logAll returns true, every query and error will be logged.
+func MonitorHook(logAll func() bool) QueryHook {
```
Comment (patrickhuie19) [inline pkg/sqlutil/monitor.go:31]: Could `MonitorHook` accept a straight `bool` arg?

**Review by patrickhuie19 — APPROVED** [review]

File: pkg/sqlutil/monitor.go (hunk 9 above)
Comment (jmank88) [reply pkg/sqlutil/monitor.go:31]: No, it is one of the few settings that can be changed at runtime, so we call this every time.

**Pull request merged**